widget/completion_entry_test.go

```diff
@@ -0,0 +1,161 @@
+}
+
+// Hide the menu on Escape key.
+func TestCompletionEntry_Escape(t *testing.T) {
+	entry := createEntry()
+	win := test.NewWindow(entry)
+	win.Resize(fyne.NewSize(500, 300))
+	defer win.Close()
+
+	entry.SetText("init")
+
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyDown})
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyEscape})
+
+	assert.False(t, entry.popupMenu.Visible())
+}
+
+// Hide the menu on rune pressed.
+func TestCompletionEntry_Rune(t *testing.T) {
+	entry := createEntry()
+	win := test.NewWindow(entry)
+	win.Resize(fyne.NewSize(500, 300))
+	defer win.Close()
+
+	entry.SetText("foobar")
+	entry.CursorColumn = 6 // place the cursor after the text
+
+	// type some chars...
+	win.Canvas().Focused().TypedRune('x')
+	win.Canvas().Focused().TypedRune('y')
+	assert.Equal(t, "foobarxy", entry.Text)
+
+	// make a move and type other char
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyDown})
+	win.Canvas().Focused().TypedRune('z')
+	assert.Equal(t, "foobarxyz", entry.Text)
+
+	assert.True(t, entry.popupMenu.Visible())
+}
+
+// Hide the menu on rune pressed.
+func TestCompletionEntry_Rotation(t *testing.T) {
+	entry := createEntry()
+	win := test.NewWindow(entry)
+	win.Resize(fyne.NewSize(500, 300))
+	defer win.Close()
+
+	entry.SetText("foobar")
+	entry.CursorColumn = 6 // place the cursor after the text
+
+	// loop one time (nb items + 1) to go back to the first element
+	for i := 0; i <= len(entry.Options); i++ {
+		win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyDown})
+	}
+
+	assert.Equal(t, 0, entry.navigableList.selected)
+
+	// Do the same in reverse order, here, onlh one time to go on the last item
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyUp})
+	assert.Equal(t, len(entry.Options)-1, entry.navigableList.selected)
+}
+
+// Test sumbission with opened completion.
+func TestCompletionEntry_OnSubmit(t *testing.T) {
+	entry := createEntry()
+	win := test.NewWindow(entry)
+	win.Resize(fyne.NewSize(500, 300))
+	defer win.Close()
+
+	entry.OnSubmitted = func(s string) {
+		entry.HideCompletion()
+		assert.True(t, entry.popupMenu.Hidden)
+	}
+	entry.OnChanged = func(s string) {
+		entry.ShowCompletion()
+	}
+
+	entry.SetText("foo")
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyReturn})
+}
+
+func TestCompletionEntry_DoubleSubmissionIssue(t *testing.T) {
+	entry := createEntry()
+	entry.SetOptions([]string{"foofoo", "bar", "baz"})
+	win := test.NewWindow(entry)
+	win.Resize(fyne.NewSize(500, 300))
+	defer win.Close()
+
+	entry.OnSubmitted = func(s string) {
+		t.Log("The entry should not submit value if Enter was press to select an option.")
+		t.Fail()
+	}
+
+	entry.OnChanged = func(s string) {
+		entry.ShowCompletion()
+	}
+	entry.SetText("foo")
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyDown})  // select foofoo
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyEnter}) // OnSubmitted should NOT be called
+}
```

Comment:
Same here:
```go
submitCalled := false
entry.OnSubmitted = func(s string) {
	submitCalled = true
}
...
win.Canvas().Focus(entry)
entry.SetText("foo")

win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyDown}) // select foofoo
assert.False(t, submitCalled)
win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyEnter}) // OnSubmitted should NOT be called
assert.False(t, submitCalled)
win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyEnter}) // OnSubmitted should be called
assert.True(t, submitCalled)
```

Reply:
Like my previous comment, this tests are made to check a known issue, the behavior to test is a "non call"

Reply:
The test name is `TestCompletionEntry_DoubleSubmissionIssue`, but you are only trying to trigger (theoretically) one submit event, shouldn't you call another one and verify if this time everything went fine? 🤔 Anyway, as I said at the beginning, the tests are ok, these are just my thoughts because I am paranoic with false positives 😅, feel free to ignore this if this has no sense.

Reply:
OK, I actually didn't understand what was the problem for you. I now understand that you want to avoid false positives.

Reply:
The problem is now that the entry loses focus after the first Enter key pressed... This doesn't happen in real life, only inside the test.

Reply:
Found... I need to "force" the focus (and I really don't know why, but that makes the job)

Reply:
> Found... I need to "force" the focus (and I really don't know why, but that makes the job)

Hmm the above code doesn't work for you? I tested it locally and the test passed 🤔

Reply:
Done https://github.com/fyne-io/fyne-x/pull/11/commits/d5226ec4ef7f385c9ef6207bfa7a68a8ce37d30c

Reply:
Hmm I don't think this line is needed:
```go
fyne.CurrentApp().Driver().CanvasForObject(c).Focus(c) // regain focus
```
It was added just because of the test?

Reply:
Yes... In test the focus is lost. And without this, the  double submission test fails.

Reply:
> Yes... In test the focus is lost. And without this, the double submission test fails.

The test passes without it. I think you forgot one line from my above suggestion:
```go
...
win.Canvas().Focus(entry) // 👈 here you are requesting focus (entry.SetText does not request focus)
entry.SetText("foo")

win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyDown}) // select foofoo
...
```

Reply:
Here, it says that there is a nil pointer dereference

Reply:
Yes, forgotten... (I should sleep before making changes...)
Fixed here: https://github.com/fyne-io/fyne-x/pull/11/commits/5eeb4701a65a9f5e3ad429ac310caa206af44d4a

Reply:
and commented here https://github.com/fyne-io/fyne-x/pull/11/commits/be3a9faf961816757cd3e680a9300f7775dd6399

Reply:
Great! Thanks :)